### PR TITLE
add cert manager smoke test to setup singleton and preload data

### PR DESCRIPTION
**Pull request by bluzarraga**
For issue 59713, added a wait for cert manager operand webhook to come ready to setup singleton and then a smoke test where it tries to create a test issuer and certificate. This was added to the end of the script. Added just the smoke test to preload data's prereqs section so the script will error out if these items cannot be created

Testing setup singleton:
- run setup singleton to fresh install cert manager
- verify script completes and the success message "Cert manager is ready." appears at the end of the script

testing preload:
- before installing cert manager, run the script
- verify the script errors out during prereqs with the message: "Failed to create test issuer. Verify cert manager is installed and ready on the cluster then re-run the preload script."
- Install cert manager
- wait for it to come ready
- run preload again, verify the preload checks complete
- Note: you do not need to have any cs installed for the preload script to complete prereq checks. It will error out after the part we are looking at (which is expected).

### CODE REVIEW

File: cp3pt0-deployment/setup_singleton.sh [hunk 1]
```diff
@@ -421,4 +422,65 @@ function get_and_validate_arguments() {
...
+spec:
+  commonName: test-certificate
+  duration: 17520h0m0s
+  issuerRef:
+    kind: Issuer
+    name: test-issuer
+  renewBefore: 720h0m0s
+  secretName: test-certificate-secret
+EOF
+    return_value_cert=$(${OC} get certificate -n $OPERATOR_NS --ignore-not-found | grep test-certificate || echo "false")
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/setup_singleton.sh:474]:
```suggestion
    return_value_cert=$(${OC} get certificate.v1.cert-manager.io -n $CERT_MANAGER_NAMESPACE --ignore-not-found | grep test-certificate || echo "false")
```

File: cp3pt0-deployment/setup_singleton.sh [hunk 2]
```diff
@@ -421,4 +422,65 @@ function get_and_validate_arguments() {
...
+    if [[ $return_value_cert == "false" ]]; then
+      ${OC} delete issuer test-issuer -n $OPERATOR_NS --ignore-not-found
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/setup_singleton.sh:476]:
```suggestion
      ${OC} delete issuer.v1.cert-manager.io test-issuer -n $CERT_MANAGER_NAMESPACE --ignore-not-found
```

File: cp3pt0-deployment/setup_singleton.sh [hunk 3]
```diff
@@ -421,4 +422,65 @@ function get_and_validate_arguments() {
...
+      error "Failed to create test certificate."
+    else
+      ${OC} delete certificate test-certificate -n $OPERATOR_NS --ignore-not-found
+      ${OC} delete issuer test-issuer -n $OPERATOR_NS --ignore-not-found
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/setup_singleton.sh:480]:
```suggestion
      ${OC} delete certificate.v1.cert-manager.io test-certificate -n $CERT_MANAGER_NAMESPACE --ignore-not-found
      ${OC} delete issuer.v1.cert-manager.io test-issuer -n $CERT_MANAGER_NAMESPACE --ignore-not-found
```

File: preload_data.sh [hunk 4]
```diff
@@ -1870,6 +1872,50 @@ function wait_for_job_complete() {
...
+  cat << EOF | ${OC} apply -f -
+apiVersion: cert-manager.io/v1
+kind: Issuer
+metadata:
+  name: test-issuer
+  namespace: $FROM_NAMESPACE
+spec:
+  selfSigned: {}
+EOF
+  return_value_issuer=$(${OC} get issuer -n $FROM_NAMESPACE --ignore-not-found | grep test-issuer || echo "false")
```
Comment (Daniel-Fan) [inline preload_data.sh:1887]:
```suggestion
  return_value_issuer=$(${OC} get issuer.v1.cert-manager.io -n $FROM_NAMESPACE --ignore-not-found | grep test-issuer || echo "false")
```

File: preload_data.sh [hunk 5]
```diff
@@ -1870,6 +1872,50 @@ function wait_for_job_complete() {
...
+spec:
+  commonName: test-certificate
+  duration: 17520h0m0s
+  issuerRef:
+    kind: Issuer
+    name: test-issuer
+  renewBefore: 720h0m0s
+  secretName: test-certificate-secret
+EOF
+    return_value_cert=$(${OC} get certificate -n $FROM_NAMESPACE --ignore-not-found | grep test-certificate || echo "false")
```
Comment (Daniel-Fan) [inline preload_data.sh:1907]:
```suggestion
    return_value_cert=$(${OC} get certificate.v1.cert-manager.io -n $FROM_NAMESPACE --ignore-not-found | grep test-certificate || echo "false")
```

File: preload_data.sh [hunk 6]
```diff
@@ -1870,6 +1872,50 @@ function wait_for_job_complete() {
...
+    if [[ $return_value_cert == "false" ]]; then
+      ${OC} delete issuer test-issuer -n $OPERATOR_NS --ignore-not-found
```
Comment (Daniel-Fan) [inline preload_data.sh:1909]:
```suggestion
      ${OC} delete issuer.v1.cert-manager.io test-issuer -n $FROM_NAMESPACE --ignore-not-found
```

File: preload_data.sh [hunk 7]
```diff
@@ -1870,6 +1872,50 @@ function wait_for_job_complete() {
...
+      error "Failed to create test certificate. Verify cert manager is installed and ready on the cluster then re-run the preload script."
+    else
+      ${OC} delete certificate test-certificate -n $FROM_NAMESPACE --ignore-not-found
+      ${OC} delete issuer test-issuer -n $FROM_NAMESPACE --ignore-not-found
```
Comment (Daniel-Fan) [inline preload_data.sh:1913]:
```suggestion
      ${OC} delete certificate.v1.cert-manager.io test-certificate -n $FROM_NAMESPACE --ignore-not-found
      ${OC} delete issuer.v1.cert-manager.io test-issuer -n $FROM_NAMESPACE --ignore-not-found
```

File: cp3pt0-deployment/setup_singleton.sh [hunk 8]
```diff
@@ -421,4 +422,65 @@ function get_and_validate_arguments() {
...
+  if [[ $return_value_issuer == "false" ]]; then
+    error "Failed to create test issuer."
+  else
+    debug1 "Creating test certificate in namespace $OPERATOR_NS."
+    cat << EOF | ${OC} apply -f -
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: test-certificate
+  namespace: $OPERATOR_NS
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/setup_singleton.sh:464]:
```suggestion
  namespace: $CERT_MANAGER_NAMESPACE
```

File: cp3pt0-deployment/setup_singleton.sh [hunk 9]
```diff
@@ -421,4 +422,65 @@ function get_and_validate_arguments() {
...
+    error "More than one cert-manager-webhook deployment exists on the cluster."
+  fi
+
+  debug1 "Creating test issuer in namespace $OPERATOR_NS."
+  cat << EOF | ${OC} apply -f -
+apiVersion: cert-manager.io/v1
+kind: Issuer
+metadata:
+  name: test-issuer
+  namespace: $OPERATOR_NS
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/setup_singleton.sh:450]:
```suggestion
  namespace: $CERT_MANAGER_NAMESPACE
```

**Review by Daniel-Fan — CHANGES_REQUESTED** [review]
Could we use full qualified resource name to distinguish `v1alpha1` version of issuer.

Also I am wondering if we should use `$CERT_MANAGER_NAMESPACE` to deploy test issuer, since `$OPERATOR_NS` is optional, and it is only used for v3.x migration.

File: cp3pt0-deployment/setup_singleton.sh [hunk 10]
```diff
@@ -421,4 +422,65 @@ function get_and_validate_arguments() {
...
+  local error_message="Timeout after ${total_time_mins} minutes waiting for pod ${name} to be running."
+  wait_for_condition "${condition}" ${retries} ${sleep_time} "${wait_message}" "${success_message}" "${error_message}"
+
+  #check no duplicate webhook pod
+  webhook_deployments=$(${OC} get deploy -A --no-headers --ignore-not-found | grep ${name} -c)
+  if [[ $webhook_deployments != "1" ]]; then
+    error "More than one cert-manager-webhook deployment exists on the cluster."
+  fi
+
+  debug1 "Creating test issuer in namespace $OPERATOR_NS."
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/setup_singleton.sh:444]:
```suggestion
  debug1 "Creating test issuer in namespace $CERT_MANAGER_NAMESPACE."
```

File: cp3pt0-deployment/setup_singleton.sh [hunk 11]
```diff
@@ -421,4 +422,65 @@ function get_and_validate_arguments() {
...
+  name: test-issuer
+  namespace: $OPERATOR_NS
+spec:
+  selfSigned: {}
+EOF
+  return_value_issuer=$(${OC} get issuer -n $OPERATOR_NS --ignore-not-found | grep test-issuer || echo "false")
+  if [[ $return_value_issuer == "false" ]]; then
+    error "Failed to create test issuer."
+  else
+    debug1 "Creating test certificate in namespace $OPERATOR_NS."
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/setup_singleton.sh:458]:
```suggestion
    debug1 "Creating test certificate in namespace $CERT_MANAGER_NAMESPACE."
```

**Review by Daniel-Fan — CHANGES_REQUESTED** [review]

**Comment by bluzarraga** [comment]
Tested on amd64 successfully after making recommended changes listed above.
Running preload without installing cert manager:
```
[root@api.admits.cp.fyre.ibm.com ibm-common-service-operator]# ./preload_data.sh --original-cs-ns cs --services-ns test-ns -v 1
No resources found in cs namespace.
[INFO] Checking cert manager readiness.
[DEBUG] Creating test issuer in namespace cs.
error: resource mapping not found for name: "test-issuer" namespace: "cs" from "STDIN": no matches for kind "Issuer" in version "cert-manager.io/v1"
ensure CRDs are installed first
error: the server doesn't have a resource type "issuer"
[✘] Failed to create test issuer. Verify cert manager is installed and ready on the cluster then re-run the preload script.
```
Output as expected

Running setup singleton:
```
...
[✔] Pod cert-manager-webhook is running.

issuer.cert-manager.io/test-issuer created
certificate.cert-manager.io/test-certificate created
certificate.cert-manager.io "test-certificate" deleted
issuer.cert-manager.io "test-issuer" deleted
[✔] Cert manager is ready.
```
Output as expected

Running preload again after cert manager is installed:
```
[root@api.admits.cp.fyre.ibm.com ibm-common-service-operator]# ./preload_data.sh --original-cs-ns cs --services-ns test-ns -v 1
[INFO] Checking cert manager readiness.
[DEBUG] Creating test issuer in namespace cs.
issuer.cert-manager.io/test-issuer created
[DEBUG] Creating test certificate in namespace cs.
certificate.cert-manager.io/test-certificate created
certificate.cert-manager.io "test-certificate" deleted
issuer.cert-manager.io "test-issuer" deleted
[✔] Cert manager is ready, preload can proceed.
...
```
Output as expected, script completed successfully after

Scripts ready for re-review

**Review by Daniel-Fan — APPROVED** [review]
/lgtm

**Pull request merged**